pkg/buildpipeline/pipeline.go

```diff
@@ -772,79 +808,123 @@ func createPipelineTasks(block *TransformedStage, blocks []TransformedStage, hea
 			if found {
-				return true, append(parallelEndTasks, provider...)
+				return true, provider
+			}
+		} else if sibling.isParallelStage() {
+			// We don't want to use a workspace from a parallel block outside of that block,
+			// but we do need to descend inwards in case stage is in that same block.
+			if stage.getEnclosing(sibling.Depth) == sibling {
+				for _, nested := range sibling.Parallel {
+					// Pick the parallel branch that has stage
+					if stage.getEnclosing(nested.Depth) == nested {
+						found, provider := findWorkspaceProvider(stage, nested)
+						if found {
+							return true, provider
+						}
+					}
+				}
 			}
-		} else if len(previousBlock.Parallel) > 0 {
-			// We don't want to use a workspace from a parallel block outside of that block.
 			// TODO: What to do about custom workspaces? Check for erroneous uses specially?
 			// Allow them if only one of the parallel tasks uses the same resource?
-
-			if previousBlock.Depth == block.Depth {
-				// TODO: We do want to record the last task in each parallel branch, though, currently hacked in as from,
-				// but will end up being used for the non-resource-based dependency ordering field.
-				parallelEndTasks = findEndTasks(previousBlock)
-			}
-		} else if previousBlock.PipelineTask != nil {
-			if *block.Stage.Options.Workspace == *previousBlock.Stage.Options.Workspace {
-				return true, append(parallelEndTasks, previousBlock.Task.Name)
+		} else if sibling.PipelineTask != nil {
+			if *sibling.Stage.Options.Workspace == *stage.Stage.Options.Workspace {
+				return true, []string{sibling.Task.Name}
 			}
 		} else {
-			// The previousBlock does not yet have a PipelineTask, so we are in a
-			// sequential block. Check the task before it.
+			// We are in a sequential block and sibling has not had its PipelineTask created.
+			// Check the task before it so we don't use a workspace of a later task.
 		}
-
+		sibling = sibling.PreviousSiblingStage
 	}
 
 	return false, nil
 }
 
 // Find the end tasks for this block, traversing down to the end blocks of any nested sequential or parallel blocks as well.
-func findEndTasks(block TransformedStage) []string {
-	if len(block.Sequential) > 0 {
-		return findEndTasks(block.Sequential[len(block.Sequential)-1])
-	} else if len(block.Parallel) > 0 {
-		var endTasks []string
+func findEndStages(block TransformedStage) []*TransformedStage {
+	if block.isSequentialStage() {
+		return findEndStages(*block.Sequential[len(block.Sequential)-1])
+	} else if block.isParallelStage() {
+		var endTasks []*TransformedStage
 		for _, pBlock := range block.Parallel {
-			endTasks = append(endTasks, findEndTasks(pBlock)...)
+			endTasks = append(endTasks, findEndStages(*pBlock)...)
 		}
 		return endTasks
-	} else if block.PipelineTask != nil {
-		return []string{block.Task.Name}
+	} else if block.Task != nil {
+		return []*TransformedStage{&block}
+	} else {
+		panic("Stage must either be sequential, parallel, or atomic!")
```

Comment:
Is `panic` the right thing to use here? Or would we be better served by returning an error?

Reply:
My thought was that if this happens, it is because we have done something very wrong constructing our data types, i.e. this is like throwing an AssertionError or IllegalStageException, and we'd like this to fail very fast so we catch it in development.

Reply:
Alright, we can revisit later if we need to.

Reply:
FWIW I went with "replace the last else if with a bare else and drop the panic, since that's what we do in other cases where we check if something is sequential, parallel, or atomic" in d6f0c8f, can revert if you think the original non-panic code was preferable.